### PR TITLE
[filedao] improve HeaderFooterByHeight() efficiency

### PR DESCRIPTION
**Pull request by dustinxie**
address #2471

### CODE REVIEW

File: blockchain/filedao/filedao.go [hunk 1]
```diff
@@ -204,6 +212,57 @@ func (fd *fileDAO) GetBlockByHeight(height uint64) (*block.Block, error) {
...
+		err error
+	)
+	if fd.v2Fd != nil {
+		if blk, err = fd.v2Fd.GetBlock(hash); err == nil {
+			return &blk.Header, nil
+		}
+	}
+
+	if fd.legacyFd != nil {
+		return fd.legacyFd.Header(hash)
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:227]: this would only get/deserialize header, instead of entire block

File: blockchain/filedao/filedao.go [hunk 2]
```diff
@@ -204,6 +212,57 @@ func (fd *fileDAO) GetBlockByHeight(height uint64) (*block.Block, error) {
...
+			blk, err := v2.GetBlockByHeight(height)
+			if err != nil {
+				return nil, err
+			}
+			return &blk.Header, nil
+		}
+	}
+
+	if fd.legacyFd != nil {
+		return fd.legacyFd.HeaderByHeight(height)
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:244]: this would only get/deserialize header, instead of entire block

**Comment by yutongp** [comment]
I don't think this addresses the issue in V2 file format

**Comment by dustinxie** [comment]
> I don't think this addresses the issue in V2 file format

right, V2 file store the entire block together, to get the header you'll need to get/deserialize the entire block.
in other words, this issue is specific only for legacy file

**Comment by yutongp** [comment]
then in that case we may have performance concern on v2 file for building bloom filter

**Comment by dustinxie** [comment]
> then in that case we may have performance concern on v2 file for building bloom filter

we did comprehensive tests to try out diff storage options for v2 file (whether or not to store multiple blocks together, how many blocks, use Snappy or Gzip, etc.), and reach the conclusion to use current v2 file implementation, which has a better performance in terms of read/write access time than legacy file

again, this issue is specific to legacy file, that HeaderByHeight() is not implemented with the best way in first place, so this PR is to correct that

**Comment by koseoyoung** [comment]
Can we make API node to only use filedao_legacy? 
I think there is an issue regarding splitting filedao and querying old data.
If a node splits the DB and has both legacy(e.g., 0-7M) and filedaov2(e.g., 7M-8M), and then query old height data(e.g., headerByHeight(6M)),  it is supposed to read from legacy filedao but it looks like reading from filedaov2 for now.

**Review by koseoyoung — APPROVED** [review]
We can either make API node to just use filedao_legacy or find another way to improve filedaoV2 headerByHeight as well. For the filedao_legacy improvement, LGTM

**Pull request merged**